lib/src/rules/prefer_final_over_var_field.dart

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.prefer_final_over_var_field;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/linter.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'Private field could be final.';
+
+const _details = r'''
+
+**DO** Prefer declaring private fields as final if they are not reassigned later
+ in the class.
+
+**BAD:**
+```
+class BadImmutable {
+  var _label = 'hola mundo! BadImmutable'; // LINT
+  var label = 'hola mundo! BadImmutable'; // OK
+}
+```
+
+**BAD:**
+```
+class MultipleMutable {
+  var _label = 'hola mundo! GoodMutable', _offender = 'mumble mumble!'; // LINT
+  var _someOther; // LINT
+
+
+  MultipleMutable() : _someOther = 5;
+
+
+  MultipleMutable(this._someOther);
+
+  void changeLabel() {
+    _label= 'hello world! GoodMutable';
+  }
+}
+```
+
+**GOOD:**
+```
+class GoodImmutable {
+  final label = 'hola mundo! BadImmutable', bla = 5; // OK
+  final _label = 'hola mundo! BadImmutable', _bla = 5; // OK
+}
+```
+
+**GOOD:**
+```
+class GoodMutable {
+  var _label = 'hola mundo! GoodMutable';
+
+  void changeLabel() {
+    _label = 'hello world! GoodMutable';
+  }
+}
+```
+''';
+
+class PreferFinalOverVarField extends LintRule {
+  _Visitor _visitor;
+
+  PreferFinalOverVarField()
+      : super(
+            name: 'prefer_final_over_var_field',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    if (node.fields.isFinal || node.fields.isConst) {
+      return;
+    }
+
+    ClassDeclaration classDeclaration =
+        node.getAncestor((a) => a is ClassDeclaration);
+    if (classDeclaration == null) {
+      return;
+    }
+
+    node.fields.childEntities.forEach((/* AstNode | Token */ entity) {
+      if (entity is! VariableDeclaration) {
+        return;
+      }
+
+      VariableDeclaration variable = entity as VariableDeclaration;
+      if (variable == null || !variable.element.isPrivate) {
+        return;
+      }
+
+      final isMutated = DartTypeUtilities
+          .traverseNodesInDFS(classDeclaration)
+          .any((n) =>
+              n is AssignmentExpression &&
+              n.leftHandSide is SimpleIdentifier &&
+              n.leftHandSide.toString() == variable.element.name);
```

Comment:
Changed.
Regarding traversing the graph I do `where` to keep only the parent nodes of `PropertyAccessorElement` which at that point should be fast to traverse.